### ISSUE TITLE
Contacts updates on the device can cause mistargeted SMS sessions

### ISSUE THREAD

**Issue by ferdnyc** [andyholmes/gnome-shell-extension-gsconnect#341]
**Describe the bug**
This is kind of a weird one. It appears that when contacts are updated on the device, it can trick GSConnect into using the wrong number for a contact.

**To Reproduce**
Steps that triggered the behavior:
1. I was  having an SMS conversation with an unknown number
2. During the conversation, I added their number to an _existing_ contact
3. After that point, GSConnect's notifications of texts mirrored from that same number started being labeled with the contact's name
4. However, the new _number_ (the one the SMS messages are actually coming from) hasn't been added in the GSConnects contacts list
5. As a result of that (presumably), when I clicked on the latest SMS notification from GSConnect, it opened an SMS window to the **wrong** number (the contact's old number, which isn't the number the text message came from)

**Expected behavior**
SMS conversation windows always open with the recipient set to the number which sent the message in the notification

**System Details (please complete the following information):**
 - GSConnect version: v15
   - Installed from: extensions.gnome.org
 - GNOME/Shell version: 3.30
 - Distro/Release: Fedora 29

**GSConnect environment (if applicable):**
 - Paired Device(s): Galaxy S6
 - KDE Connect app version: latest beta from Google play
 - Messaging App: Android Messages

**Comment by andyholmes** [comment]
Did the original contact, before adding the new number, only have a single phone number? If so I think I can fix this easily as a part of #332.

Otherwise, we might have to drop opening directly to a contact from the notification and just open a blank window. I'm not sure how fast contacts are synced from the phone, but if I recall syncing through libfolks takes quite some time; possibly even a restart of gnome-online-accounts.

**The longer, depressing reality**

SMS notifications don't actually carry any contact information, per se; they only have a title. The GNotification action uses that as a "hint", searching first for a phone number or if the "hint" contains no digits, searches for an exact name match. If an exact name match with a single phone number is found, we just assume that it's the right contact/number.

All of this is (of course) complicated by the fact that two or more "contacts" can very much have both the same name *and* phone number, both in real life *and* by way of differing account IDs with the same information (or all four of those :confused:).

**Comment by ferdnyc** [comment]
> Did the original contact, before adding the new number, only have a single phone number? If so I think I can fix this easily as a part of #332.

Nope, increased from 2 to 3 with the addition. (So, only 2 were showing in the GSConnect contacts list.) But, it sounds from your correction like that wouldn't actually have mattered anyway?

> SMS notifications don't actually carry any contact information, per se; they only have a title. The GNotification action uses that as a "hint", searching first for a phone number or if the "hint" contains no digits, searches for an exact name match.

So, the title is _either_ a phone number (for unknown contacts), or _only_ a contact name (for known contacts)? Eugggh, that's _**horrible**_! I was hoping that at least the phone# was always sent through. Then, if a notification came through with a _known_ name, but attached to an _unknown_ number (or, unassociated with that name), the notification could be treated as an implicit contacts-list update, and the number just added in. But if there's no phone# at all, for known contacts... how can _any_ SMS notification be responded to? The number is kind of an essential component.

(How does the new in-notification reply feature work, does it actually send the reply back to the app that posted the notification in a way that'll cause it to target the correct number? Or does this really create the same issue for that response path?)

**Comment by andyholmes** [comment]
> But, it sounds from your correction like that wouldn't actually have mattered anyway?

Correct, and it was kind of sketchy anyways, due to the described name conflicts.

> So, the title is either a phone number (for unknown contacts), or only a contact name

It's actually just the notification title verbatim. Essentially the `kdeconnect.notification` packet is a mirror of the attributes and values of libnotify (which makes sense) with an additional "internal id" from Android that's pretty poorly defined.

> How does the new in-notification reply feature work

Basically each notification packet that's repliable will come with a UUID we can respond with a bit of text.

```js
{
  "id": 1541767166010,
  "type": "kdeconnect.notification",
  "body": {
    // md5 hash of icon; only used for avoiding re-downloads - not confirmed
    "payloadHash": "ec6586b6d0c86cafae7808258924432a",
    // This is a UUID kdeconnect generates (changes each time we're sent notifications)
    "requestReplyId": "f99ff93d-247e-4892-9d47-5895203b6532",
    //     ?| Android app-id                  |?| seems to vary :???: app notif no? ---v|UID/GID?
    "id": "0|com.google.android.apps.messaging|0|com.google.android.apps.messaging:sms:1|10079",
    // same as libnotify
    "appName": "Messages",
    // We can respond with 'id' above and 'cancel: true' to close it
    "isClearable": true,
    // backwards compatibility; always equivalent to 'title: text'
    "ticker": "My contact: Hello",
    // same as libnotify
    "title": "My contact",
    // same as libnotify's "body"
    "text": "Hello",
    // "posted" time on Android, not necessarily received time (eg. changes if phone is reset)
    "time": "1541767165992"
  },
  "payloadSize": 8082,
  "payloadTransferInfo": {
    "port": 1750
  }
}
```
